### PR TITLE
[MNG-7619] Reverse Dependency Tree

### PR DESCRIPTION
**Pull request by cstamas**
Adds Maven feature that is able to explain why an artifact is present in local repository.

Usable for diagnosing resolution issues.

In local repository, for each artifact it records `.tracking` folder, containing farthest artifact that got to this artifact, and the list of graph nodes that lead to it.

Note: this is based on by @grgrzybek proposal and reuses some code he provided. See https://github.com/apache/maven-resolver/pull/182

---

https://issues.apache.org/jira/browse/MNG-7619

### CODE REVIEW

**Comment by cstamas** [comment]
Excercise: build this PR (will produce Maven 3.9.0-SNAPSHOT distro), then use same distro to build itself (or just maven.3.9.x, does not matter) using these:
```
mvn clean install -Dmaven.repo.local=local -Dmaven.repo.local.recordReverseTree -Drat.skip
```
(rat is dumb to inspect "local" where local repository is and fail, so we skip it).

Q1: why is maven-core 2.2.1 in local repo?
```
[cstamas@infinity maven (maven-3.9.x +%)]$ tree -L 1 local/org/apache/maven/maven-core/2.2.1/.tracking/
local/org/apache/maven/maven-core/2.2.1/.tracking/
├── org.apache.maven.plugins_maven-remote-resources-plugin_jar_1.7.0
├── org.codehaus.mojo_build-helper-maven-plugin_jar_1.12
└── org.codehaus.mojo_buildnumber-maven-plugin_jar_1.4

0 directories, 3 files
[cstamas@infinity maven (maven-3.9.x +%)]$ cat local/org/apache/maven/maven-core/2.2.1/.tracking/org.codehaus.mojo_build-helper-maven-plugin_jar_1.12 
org.apache.maven:maven-core:jar:2.2.1 (compile) (plugin)
  org.codehaus.mojo:build-helper-maven-plugin:jar:1.12 () (plugin)
[cstamas@infinity maven (maven-3.9.x +%)]$
```

File: maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.eclipse.aether.graph.DependencyNode;
+
+/**
+ * A class building reverse tree using {@link CollectStepData} trace data provided in {@link RepositoryEvent}
+ * events fired during collection.
+ *
+ * @since 3.9.0
+ */
+class ReverseTreeRepositoryListener extends AbstractRepositoryListener {
+    private static final String EOL = System.lineSeparator();
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java:45]: LS, not EOL

File: maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            String trackingData = node + " (" + collectStepTrace.getContext() + ")" + EOL;
+            String indent = "";
+            ListIterator<DependencyNode> iter = collectStepTrace
+                    .getPath()
+                    .listIterator(collectStepTrace.getPath().size());
+            while (iter.hasPrevious()) {
+                DependencyNode curr = iter.previous();
+                indent += "  ";
+                trackingData += indent + curr + " (" + collectStepTrace.getContext() + ")" + EOL;
+            }
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java:85]: For cases with a loop a string builder is advised.

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by cstamas** [comment]
Note: this is based on by @grgrzybek proposal and reuses some code he provided. See https://github.com/apache/maven-resolver/pull/182

**Comment by michael-o** [comment]
Can we leave out these empty parens: `org.apache.maven.plugins:maven-dependency-plugin:jar:3.3.0 () (plugin)`?  From a user PoV it looks like a bug.

**Comment by cstamas** [comment]
Thats org.eclipse.aether.graph.Dependency#toString, we could fix it, but given this is "advanced" feature, I don't think is something would block this?

**Comment by michael-o** [comment]
> Thats org.eclipse.aether.graph.Dependency#toString, we could fix it, but given this is "advanced" feature, I don't think is something would block this?

Agree, we need to spin off the issue there. Here, no change is required.

**Comment by laeubi** [comment]
Not sure if this was already mentioned, but instead of "bake this in" would it not be better to have a way for (core-)plugins to hook in a `RepositoryListener` or even on the project level as we already to for `AbstractMavenLifeCycleListeners` / `WorkspaceReaders`? The this can become part of m-dependency-p for example.

I also wonder if it is even required to record the data at all, should it not be sufficient to first load the pom(s), then generate a list of used plugins and then resolve their dependencies and then do something similar to `dependency:tree` (maybe plugin-dependency-tree)? To make it more convenient then one might pass a `-DdependencyReverse=<GAV>` to have a reverse output rooted at the given GAV when build finished.

**Comment by cstamas** [comment]
@laeubi while I agree with ideas you wrote, I am still for this PR to go in, here is why: 
This PR is (nearly) trivial, but gives you powerful tool (for investigation, answering several questions), is registered early (earliest possible), and does not prevent anything of those you wrote (a plugin). Moreover, by being earliest present, and present during whole session of a real build, it covers _everything_ (even some plugin that may dynamically resolve). Not easily done with something like build extension or even plugin (they are all too late to cover everything).

Also, none of these prevents anything of those things you wrote, this is totally orthogonal. I personally see this like `-X`, a handy switch seldom used by users, but when one needs it, is very very handy.

File: maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+class ReverseTreeRepositoryListener extends AbstractRepositoryListener {
+    @Override
+    public void artifactResolved(RepositoryEvent event) {
+        requireNonNull(event, "event cannot be null");
+
+        if (!event.getArtifact()
+                .getFile()
+                .getPath()
+                .startsWith(event.getSession().getLocalRepository().getBasedir().getPath())) {
+            return; // reactor artifact
```
Comment (pzygielo) [inline maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java:54]: Could this condition be extracted to method IDK, `isReactorArtifact(event)`?
The comment then removed?
The method tested?

File: maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        RequestTrace trace = event.getTrace();
+        CollectStepData collectStepTrace = null;
```
Comment (pzygielo) [inline maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java:57]: Could this variable be moved one line up, and the loop then extracted to (tested) class method?

**Comment by cstamas** [comment]
Um, unsure why GH says "removed request" from @pzygielo , I actually asked both @michael-o and @pzygielo for repeated review...

File: maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java (hunk 4 above)
Comment (pzygielo) [reply maven-core/src/main/java/org/apache/maven/internal/aether/ReverseTreeRepositoryListener.java]: Class method (static) would need no instance for test. Like it doesn't to do its job.

**Review by michael-o — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #8833

**Pull request merged**